ffi_internals/src/consumer/consumer_fn.rs

```diff
@@ -0,0 +1,154 @@
+        let (return_conversion, close_conversion, return_sig) =
+            if let Some(return_type) = &self.return_type {
+                let ty = return_type.consumer_type(None);
+                (
+                    if return_type.is_result {
+                        "handle(result: ".to_string()
+                    } else {
+                        format!("{}.fromRust(", ty)
+                    },
+                    ")".to_string(),
+                    if return_type.is_result {
+                        format!("-> Result<{}, RustError>", ty)
+                    } else {
+                        format!("-> {}", ty)
+                    },
+                )
+            } else {
+                (String::new(), String::new(), String::new())
+            };
+        format!(
+            r#"
+    {static_keyword} func {consumer_fn_name}({consumer_parameters}) {return_sig} {{
+        {return_conversion}{ffi_fn_name}({ffi_parameters}){close_conversion}
+    }}
+            "#,
+            static_keyword = static_keyword,
+            consumer_fn_name = self.fn_name.to_string().to_mixed_case(),
+            consumer_parameters = self.consumer_parameters(),
+            return_sig = return_sig,
+            return_conversion = return_conversion,
+            ffi_fn_name = self.ffi_fn_name(module_name).to_string(),
+            ffi_parameters = self.ffi_calling_arguments(),
+            close_conversion = close_conversion,
+        )
+    }
+
+    /// Generates the contents of a consumer extension for this function, extending the original
+    /// type with the behaviors described by `self`. This is primarily for use with
+    /// `ffi_derive::expose_fn`, where we want to generate an FFI and consumer for a standalone
+    /// function.
+    ///
+    #[must_use]
+    pub fn generate_consumer_extension(
+        &self,
+        header: &str,
+        consumer_type: &str,
+        module_name: &Ident,
+        imports: Option<&str>,
+    ) -> String {
+        // Include the keyword `static` if this function doesn't take a receiver.
+        let static_keyword = if self.receiver == FnReceiver::None {
+            "static"
+        } else {
+            ""
+        };
+        let (return_conversion, close_conversion, return_sig) =
+            if let Some(return_type) = &self.return_type {
+                let ty = return_type.consumer_type(None);
+                (
+                    if return_type.is_result {
+                        "handle(result: ".to_string()
+                    } else {
+                        format!("{}.fromRust(", ty)
+                    },
+                    ")".to_string(),
+                    if return_type.is_result {
+                        format!("-> Result<{}, RustError>", ty)
+                    } else {
+                        format!("-> {}", ty)
+                    },
+                )
+            } else {
+                (String::new(), String::new(), String::new())
+            };
```

Comment:
Is this the same as in the previous function? Maybe DRY up?

Reply:
Oof thanks, moved to an associated fn on `TypeFFI` so they can both call it there.